### PR TITLE
[Android] If the keyboard view is not ready, calling javascript functions will crash

### LINKED ISSUE #1884: [Android] Keyboard stability issues
**Describe the bug**
Current alpha versions of the Keyman for Android app are prone to slight keyboard instability.  Sometimes the keyboards will crash and reset back to SIL EuroLatin and other times they'll crash outright, leaving a blank (black) WebView and no ability to swap keyboards.  (This is especially damaging when in the system keyboard.)

**To Reproduce**
Part of the problem is that we currently lack an exact reproduction for this, though there are situations where it is prone to occur.  Note [this comment](https://github.com/keymanapp/keyman/pull/1881#pullrequestreview-264918806) on #1881 for examples.

**Screenshots**
See https://github.com/keymanapp/keyman/pull/1881#pullrequestreview-264918806.

---

One particular standout element of the link above:

```
D/KMEA: Keyman JS Log: Line 1, file:///data/user/0/com.tavultesoft.kmapro.debug/app_data/keyboard.html:Uncaught ReferenceError: updateKMText is not defined
D/FA: Logging event (FE): kmw_console_error, Bundle[{keyboardVersion=1.8.1, cm_lineNumber=1, keyboardID=sil_euro_latin, firebase_event_origin(_o)=app, firebase_screen_class(_sc)=MainActivity, firebase_screen_id(_si)=-1257932794863401442, cm_sourceID=file:///data/user/0/com.tavultesoft.kmapro.debug/app_data/keyboard.html, keyboardName=EuroLatin (SIL) Keyboard, keyboardType=INAPP, cm_message=Uncaught ReferenceError: updateKMText is not defined, packageID=cloud}]
D/KMEA: Keyman JS Log: Line 1, file:///data/user/0/com.tavultesoft.kmapro.debug/app_data/keyboard.html:Uncaught ReferenceError: updateKMSelectionRange is not defined
I/Choreographer: Skipped 83 frames!  The application may be doing too much work on its main thread.
E/eglCodecCommon: glUtilsParamSize: unknow param 0x000085b5
   glUtilsParamSize: unknow param 0x000085b5
D/FA: Logging event (FE): kmw_console_error, Bundle[{keyboardVersion=1.8.1, cm_lineNumber=1, keyboardID=sil_euro_latin, firebase_event_origin(_o)=app, firebase_screen_class(_sc)=MainActivity, firebase_screen_id(_si)=-1257932794863401442, cm_sourceID=file:///data/user/0/com.tavultesoft.kmapro.debug/app_data/keyboard.html, keyboardName=EuroLatin (SIL) Keyboard, keyboardType=INAPP, cm_message=Uncaught ReferenceError: updateKMSelectionRange is not defined, packageID=cloud}]
```

To focus more closely...

```
keyboard.html:Uncaught ReferenceError: updateKMText is not defined
keyboard.html:Uncaught ReferenceError: updateKMSelectionRange is not defined
```

The weird thing - keyboard.html is where those two functions _are_ defined.  Somehow they're getting called either during a page refresh or before these functions' definitions have been processed during the page load.

Either way, the functions themselves are called by KMEA - but since the definitions aren't available at the time, a JavaScript error is generated.  With KMEA's current construction, JavaScript errors always trigger keyboard resets, causing at least one of the problems listed above.  Might it be capable of causing infinite reset loops and/or blank keyboard views as well?

---

In internal discussions, we've noted that we may need to manage implement a JavaScript call 'queue'  to ensure KMEA calls to the KMW engine can only occur during valid states.  There's a good chance this may also help improve the smoothness of keyboard loads in general.  The main question:  is the amount of work involved in implementing this light enough to make a reasonable addition to 12.0's beta, or will it need to be withheld for 13.0 work instead?

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #1884. 

This is not reproducible on all devices, but on some devices it seems that Javascript functions are called even when the page hasn't finished loading. So this fix is multi-part:

1. If any Javascript call is made from the engine to the web frame before the page finishes loading, we cache it and issue it after the load completes.
2. The `WebView.onPageFinished` event occurs too soon after the page load, and script has not necessarily finished executing. This can cause the Javascript calls to fail. So we now issue `'pageLoaded'` event when the web page calls its own `window.onloaded` event, and only mark the page as ready then.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/158129:id/kvk/index.html)

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 1]
```diff
@@ -1342,11 +1342,13 @@ public void onPageFinished(WebView view, String url) {
           public void run() {
             SharedPreferences prefs = context.getSharedPreferences(context.getString(R.string.kma_prefs_name), Context.MODE_PRIVATE);
             if (prefs.getBoolean(KMManager.KMKey_ShouldShowHelpBubble, true)) {
-              InAppKeyboard.loadUrl("javascript:showHelpBubble()");
+              InAppKeyboard.loadJavascript("showHelpBubble()");
             }
           }
         }, 2000);
 
+        InAppKeyboard.callJavascriptAfterLoad();
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: No `SystemKeyboard` version yet?

Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: Nup, just testing inapp to start with

Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: Ah, actually, my bad. The rest should be in place already.

Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: Fixed

**Comment by mcdurdin** [comment]
This fixes point 7 in #1904.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 2]
```diff
@@ -1575,6 +1575,8 @@ public void run() {
         }, 2000);
 
         KeyboardEventHandler.notifyListeners(KMTextView.kbEventListeners, KeyboardType.KEYBOARD_TYPE_SYSTEM, EventType.KEYBOARD_LOADED, null);
+
+        SystemKeyboard.callJavascriptAfterLoad();
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: 👍

**Pull request merged**